File: 876-hand-of-straights/hand-of-straights.java
class Solution {
    public boolean isNStraightHand(int[] hand, int groupSize) {
        TreeMap<Integer,Integer> mp=new TreeMap<>();
        for(int i:hand){
            mp.put(i,mp.getOrDefault(i,0)+1);
        }

        while(mp.size()>0){
            Set s=mp.keySet();
            Iterator<Integer> it=s.iterator();
            int prev=-1;
            int n=groupSize;
            while(n>0 && it.hasNext()){
                n--;
                int val=it.next();
                if(val-prev>1&&prev!=-1) return false;
                if(mp.get(val)==1) it.remove();
                else{
                    mp.put(val,mp.get(val)-1);
                }
                prev=val;
            }
            if(n>0) return false;
        }
        return true;
    }
}

File: 876-hand-of-straights/hand-of-straights.cpp
class Solution {
public:
    bool isNStraightHand(vector<int>& hand, int groupSize) {
        ios_base::sync_with_stdio(0);
        multiset<int> ms;
        for(int i:hand) 
            ms.insert(i);
        
        while(ms.size()){
            int val=*ms.begin();
            ms.erase(ms.find(val));

            for(int i=1;i<groupSize;i++){
                if(ms.find(val+i)==ms.end()) return false;
                ms.erase(ms.find(val+i));
            }
        }
        return 1;
    }
};